src/queries/role.ts

```diff
@@ -28,22 +29,73 @@ const matchAccessibleNameIfNeeded = (
   );
 };
 
+const accessibilityStates = [
+  'disabled',
+  'selected',
+  'checked',
+  'busy',
+  'expanded',
+] as const;
+
 const queryAllByRole = (
   instance: ReactTestInstance
 ): ((role: TextMatch, options?: ByRoleOptions) => Array<ReactTestInstance>) =>
   function queryAllByRoleFn(role, options) {
-    return instance.findAll(
-      (node) =>
+    return instance.findAll((node) => {
+      const matchRole =
         typeof node.type === 'string' &&
-        matchStringProp(node.props.accessibilityRole, role) &&
-        matchAccessibleNameIfNeeded(node, options?.name)
-    );
+        matchStringProp(node.props.accessibilityRole, role);
+
+      if (!matchRole) return false;
+
+      if (options?.name) {
+        if (!matchAccessibleNameIfNeeded(node, options.name)) {
+          return false;
+        }
+      }
+
+      return accessibilityStates.every((accessibilityState) => {
```

Comment:
Can we extract the last part into `matchAccessibilityStateIfNeeded`?

Reply:
Is there an option to use the same predict for checking a11yState in both `*ByRole` and `*ByA11yState` queries?

Reply:
Not as is, because `ByA11yState` checks for exact match (options == state), while here we want all options to match state, but it's fine skipping some. And we do match `false` and `undefined`.

Reply:
I don't think this is actually the case. Looks like `*ByA11yState` is only checking the expected props and ignoring others:

```ts
  function queryAllByA11yStateFn(state) {
    return instance.findAll(
      (node) =>
        typeof node.type === 'string' &&
        matchObjectProp(node.props.accessibilityState, state)
    );
  };
```

```ts
export function matchObjectProp<T extends Record<string, unknown>>(
  prop: T | undefined,
  matcher: T
): boolean {
  if (!prop || Object.keys(matcher).length === 0) {
    return false;
  }

  return (
    Object.keys(prop).length !== 0 &&
    Object.keys(matcher).every((key) => prop[key] === matcher[key])
  );
}
```

Reply:
The only difference I can spot is that `*ByRole` `state` checks default missing state options as `false` while `*ByA11yState` treats `undefined` and `false` as different values. I think both of these queries should have the behaviour you proposed for `*ByRole` query, as we want to model the underlying RN behaviour, which treats `undefined` the same as `false`. Additionally such discrepancy will confuse the users.

Therefore, I think we should fix by `*ByA11yState` query handling. I would even consider that rather to be a bug fix to that query, rather than a change in behaviour, but we can consider both options.

Reply:
Yep you're right about skipping non relevant props, but it doesn't check implicit state, so all the tests in the form of `returns an implicitly enabled element` fail.

Reply:
Therefore, I think we should change `*ByA11yState` matching to include `default` states, because the underlying RN behaviour is to treat `undefined` in the same way as `false`. Otherwise that query contradicts platform behaviour, while our goal is to simulate RN behaviour the best we can.

Reply:
I've logged #1165 in order to fix `*ByA11yState` query behavior